File: top-interview-150/0202-happy-number/0202-happy-number.cpp
class Solution {
public:
    bool isHappy(int n) {
        unordered_map<int, bool> seen;
        while (n != 1) {
            if (seen.count(n) > 0) {
                return false;
            }
            seen[n] = true;
            
            int sum = 0;
            while (n > 0) {
                sum += pow(n % 10, 2) ;
                n /= 10;
            }
            n = sum;
        }
        
        return true;
    }
};

File: top-interview-150/0202-happy-number/0202-happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        def calc_sum_squares(n):
            return sum([int(digit) ** 2 for digit in str(n)])

        tortoise = calc_sum_squares(n)
        hare = calc_sum_squares(calc_sum_squares(n))
        while tortoise != hare:
            tortoise = calc_sum_squares(tortoise)
            hare = calc_sum_squares(calc_sum_squares(hare))
            
        return hare == 1
        